### ISSUE TITLE
Circuit2Verilog - Splitter Bug? 

### ISSUE THREAD

**Issue by satu0king** [CircuitVerse/CircuitVerse#1825]
Splitter Circuit
![image](https://user-images.githubusercontent.com/2092958/96707317-55894580-13b5-11eb-8912-86dd540160b7.png)

Combiner Circuit
![image](https://user-images.githubusercontent.com/2092958/96707295-4e623780-13b5-11eb-924e-5dd1ed960eac.png)

@drjyeh please check this. Is this the intended output? 

```

module Splitter(out_0, out_1, inp_0);
  output out_0,  out_1;
  input [1:0] inp_0;
  
  assign out_1 = inp_0[1];
  assign out_0 = inp_0[0];
endmodule

module Combiner(out_0, inp_0, inp_1);
  output [1:0] out_0;
  input inp_0, inp_1;
  wire [1:0] Splitter_0_inp;
  assign Splitter_0_inp = {inp_1,inp_0};
  assign out_0 = Splitter_0_inp;
endmodule
```

**Comment by satu0king** [comment]
I think this is the intended circuit, can you give a circuit example where this is failing?

**Comment by drjyeh** [comment]
Hi,

Please see a slightly more complicated circuit:


https://circuitverse.org/users/37205/projects/splitter-test <https://circuitverse.org/users/37205/projects/splitter-test>

The code generated is follows:

module Main(out_0, clk, rst, inp_1, inp_2);
  output out_0;
  input clk, rst, inp_1, inp_2;

  wire and_0_out, clk, Splitter_1_inp[1], Splitter_1_inp[0];
  wire [1:0] Splitter_1_inp, state_Q;

  assign out_0 = and_0_out;
  
  DflipFlop #(2) state(state_Q, state_Q_inv, clk, Splitter_1_inp, rst, , );
  and and_0(and_0_out, state_Q[0], state_Q[1]);
  not not_1(Splitter_1_inp[1], inp_2);
  not not_0(Splitter_1_inp[0], inp_1);
endmodule

module DflipFlop(q, q_inv, clk, d, a_rst, pre, en);
  parameter WIDTH = 1;
  output reg [WIDTH-1:0] q, q_inv;
  input clk, a_rst, pre, en;
  input [WIDTH-1:0] d;
  
  always @ (posedge clk or posedge a_rst)
    if (a_rst) begin
      q <= 'b0;
      q_inv <= 'b1;
    end else if (en == 0) ;
    else begin
      q <= d;
      q_inv <= ~d;
    end
endmodule

This is incorrect, the RED parts need be suppressed.

Thanks,
James 
---
James H-J Yeh, Ph.D.
Associate Professor, Engineering and Computer Science
Azusa Pacific University

**Comment by satu0king** [comment]
@drjyeh I am getting the following output, have you updated your code from the latest branch?
This is how it looks for me
```
module Main(out_0, clk, rst, inp_1, inp_2);
  output out_0;
  input rst, inp_1, inp_2, clk;
  wire and_0_out, not_1_out, not_0_out;
  wire [1:0] state_Q, Splitter_1_inp;
  DflipFlop #(2) state(state_Q, state_Q_inv, clk, Splitter_1_inp, rst, , );
  and and_0(and_0_out, state_Q[0], state_Q[1]);
  assign out_0 = and_0_out;
  not not_1(not_1_out, inp_2);
  assign Splitter_1_inp = {not_1_out,not_0_out};
  not not_0(not_0_out, inp_1);
endmodule
module DflipFlop(q, q_inv, clk, d, a_rst, pre, en);
  parameter WIDTH = 1;
  output reg [WIDTH-1:0] q, q_inv;
  input clk, a_rst, pre, en;
  input [WIDTH-1:0] d;
  always @ (posedge clk or posedge a_rst)
    if (a_rst) begin
      q <= 'b0;
      q_inv <= 'b1;
    end else if (en == 0) ;
    else begin
      q <= d;
      q_inv <= ~d;
    end
endmodule
```

**Comment by drjyeh** [comment]
Hi,

This has been fix with the latest commit.

The generated code is:

module Main(out_0, clk, rst, inp_1, inp_2);
  output out_0;
  input clk, rst, inp_1, inp_2;

  wire and_0_out, clk;
  wire [1:0] Splitter_1_inp, state_Q;

  
  assign out_0 = and_0_out;
  DflipFlop #(2) state(state_Q, state_Q_inv, clk, Splitter_1_inp, rst, , );
  and and_0(and_0_out, state_Q[0], state_Q[1]);
  not not_1(Splitter_1_inp[1], inp_2);
  not not_0(Splitter_1_inp[0], inp_1);
endmodule

module DflipFlop(q, q_inv, clk, d, a_rst, pre, en);
  parameter WIDTH = 1;
  output reg [WIDTH-1:0] q, q_inv;
  input clk, a_rst, pre, en;
  input [WIDTH-1:0] d;
  
  always @ (posedge clk or posedge a_rst)
    if (a_rst) begin
      q <= 'b0;
      q_inv <= 'b1;
    end else if (en == 0) ;
    else begin
      q <= d;
      q_inv <= ~d;
    end
endmodule

Thanks,
James
---
James H-J Yeh, Ph.D.
Associate Professor, Engineering and Computer Science
Azusa Pacific University

**Comment by drjyeh** [comment]
Hi,

Yes but for the other circuit the previous commits does not work.
https://circuitverse.org/users/37205/projects/splitter-test <https://circuitverse.org/users/37205/projects/splitter-test>

I have just provide a NEW commit that fixes the problem by creating a new noGenerate list.

Thanks,
James 
---
James H-J Yeh, Ph.D.
Associate Professor, Engineering and Computer Science
Azusa Pacific University

**Comment by drjyeh** [comment]
It is my hope that with this way, I can merge all the changes you made with PR1814 to work with splitters.

Would you please resolve the conflict between my newest commit and your newest branch, and I will test and try to make everything work.

Thanks,
James 
---
James H-J Yeh, Ph.D.
Associate Professor, Engineering and Computer Science
Azusa Pacific University

**Comment by satu0king** [comment]
@drjyeh I have actually merged everything you sent and its all working together locally on my laptop. Can you check slack?

**Comment by drjyeh** [comment]
Hi,

Thanks. That is great.

I have made a few more modifications that potentially solved the Splitter problem for many situations.  Also I merged most of the code.

The previous code would fail on the following:
https://circuitverse.org/users/37205/projects/splitter-test <https://circuitverse.org/users/37205/projects/splitter-test>

The new commits would fix this problem.

Thanks,
James

---
James H-J Yeh, Ph.D.
Associate Professor, Engineering and Computer Science
Azusa Pacific University
